pallets/order-book/src/tests.rs

```diff
@@ -380,6 +389,325 @@ fn fill_order_full_works() {
+						fulfillment_amount: partial_buy_amount,
+						currency_in: order.asset_in_id,
+						currency_out: order.asset_out_id,
+						sell_rate_limit: order.max_sell_rate,
+					})
+				);
+			});
+		}
+	}
+
+	#[test]
+	fn fill_order_partial_with_full_amount_works() {
+		new_test_ext().execute_with(|| {
+			let buy_amount = 100 * CURRENCY_AUSD_DECIMALS;
+			let min_fulfillment_amount = 1 * CURRENCY_AUSD_DECIMALS;
+			let sell_ratio = FixedU128::checked_from_rational(3u32, 2u32).unwrap();
+
+			assert_ok!(OrderBook::place_order(
+				ACCOUNT_0,
+				DEV_AUSD_CURRENCY_ID,
+				DEV_USDT_CURRENCY_ID,
+				buy_amount,
+				sell_ratio,
+				min_fulfillment_amount,
+			));
+
+			let (order_id, order) = get_account_orders(ACCOUNT_0).unwrap()[0];
+
+			assert_ok!(OrderBook::fill_order_partial(
+				RuntimeOrigin::signed(ACCOUNT_1),
+				order_id,
+				buy_amount,
+			));
+
+			assert_eq!(
+				AssetPairOrders::<Runtime>::get(DEV_AUSD_CURRENCY_ID, DEV_USDT_CURRENCY_ID),
+				vec![]
+			);
+
+			let max_sell_amount = OrderBook::convert_with_ratio(
+				order.asset_in_id,
+				order.asset_out_id,
+				order.max_sell_rate,
+				buy_amount,
+			)
+			.unwrap();
+
+			assert_eq!(
+				System::events()[2].event,
+				RuntimeEvent::OrmlTokens(orml_tokens::Event::Unreserved {
+					currency_id: DEV_USDT_CURRENCY_ID,
+					who: ACCOUNT_0,
+					amount: max_sell_amount
+				})
+			);
+			assert_eq!(
+				System::events()[3].event,
+				RuntimeEvent::OrmlTokens(orml_tokens::Event::Transfer {
+					currency_id: DEV_AUSD_CURRENCY_ID,
+					to: ACCOUNT_0,
+					from: ACCOUNT_1,
+					amount: buy_amount
+				})
+			);
+			assert_eq!(
+				System::events()[4].event,
+				RuntimeEvent::OrmlTokens(orml_tokens::Event::Transfer {
+					currency_id: DEV_USDT_CURRENCY_ID,
+					to: ACCOUNT_1,
+					from: ACCOUNT_0,
+					amount: max_sell_amount
+				})
+			);
+			assert_eq!(
+				System::events()[5].event,
+				RuntimeEvent::OrderBook(Event::OrderFulfillment {
+					order_id,
+					placing_account: order.placing_account,
+					fulfilling_account: ACCOUNT_1,
+					partial_fulfillment: false,
+					fulfillment_amount: buy_amount,
+					currency_in: order.asset_in_id,
+					currency_out: order.asset_out_id,
+					sell_rate_limit: order.max_sell_rate,
+				})
+			);
+		});
+	}
+
+	#[test]
+	fn fill_order_partial_bad_origin() {
+		new_test_ext().execute_with(|| {
+			let buy_amount = 100 * CURRENCY_AUSD_DECIMALS;
+			let min_fulfillment_amount = 10 * CURRENCY_AUSD_DECIMALS;
+			let sell_ratio = FixedU128::checked_from_rational(3u32, 2u32).unwrap();
+
+			assert_ok!(OrderBook::place_order(
+				ACCOUNT_0,
+				DEV_AUSD_CURRENCY_ID,
+				DEV_USDT_CURRENCY_ID,
+				buy_amount,
+				sell_ratio,
+				min_fulfillment_amount,
+			));
+
+			let (order_id, _) = get_account_orders(ACCOUNT_0).unwrap()[0];
+
+			assert_noop!(
+				OrderBook::fill_order_partial(
+					RawOrigin::None.into(),
+					order_id,
+					min_fulfillment_amount,
+				),
+				BadOrigin
+			);
+		});
+	}
```

Comment:
🤘 Thanks for adding this test as well, very thorough! 

Nit: We don't need to set up an order, as the origin check is performed first
```suggestion
	#[test]
	fn fill_order_partial_bad_origin() {
		new_test_ext().execute_with(|| {
			assert_noop!(
				OrderBook::fill_order_partial(
					RawOrigin::None.into(),
					Default::default(),
					Default::default(),
				),
				BadOrigin
			);
		});
	}
```

Reply:
Ah, yes that makes sense. Copy/pasta error.